File: platforms/react-native/modules/@shopify/checkout-kit-react-native/android/src/main/java/com/shopify/reactnative/checkoutkit/DispatchHandle.java
package com.shopify.reactnative.checkoutkit;

import androidx.annotation.NonNull;

/** Gates events after a checkout presentation ends. */
public class DispatchHandle implements DispatchCallback {
  private final DispatchCallback downstream;
  private boolean released = false;

  public DispatchHandle(@NonNull DispatchCallback downstream) {
    this.downstream = downstream;
  }

  @Override
  public synchronized void invoke(String json) {
    if (!released) {
      downstream.invoke(json);
    }
  }

  public synchronized void release() {
    released = true;
  }

  public synchronized boolean isReleased() {
    return released;
  }
}

File: platforms/react-native/modules/@shopify/checkout-kit-react-native/android/src/main/java/com/shopify/reactnative/checkoutkit/DispatchHandle.kt
package com.shopify.reactnative.checkoutkit

/**
 * Shared per-presentation dispatch handle.
 *
 * SDK lifecycle events and protocol events both invoke the same handle. Terminal lifecycle events
 * release it so subsequent protocol emissions are dropped, matching the iOS
 * pendingDispatchCallback lifecycle.
 */
class DispatchHandle(
    private val downstream: DispatchCallback,
) : DispatchCallback {
    private var released: Boolean = false

    @Synchronized
    override fun invoke(json: String) {
        if (!released) {
            downstream.invoke(json)
        }
    }

    @Synchronized
    fun release() {
        released = true
    }

    @Synchronized
    fun isReleased(): Boolean = released
}
